crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1024,59 +1041,58 @@ impl<'db> TypeInferenceBuilder<'db> {
                 format_import_from_module(level.get(), module),
                 self.file.path(self.db),
             );
-            self.relative_module_name(module, level)
+            self.relative_module_name(module, level, import_from)
         } else {
             tracing::trace!(
                 "Resolving imported object '{}' from module '{}'",
                 alias.name,
                 format_import_from_module(*level, module),
             );
-            module
-                .and_then(ModuleName::new)
-                .ok_or(ModuleResolutionError::InvalidSyntax)
+            module.and_then(ModuleName::new)
         };
 
-        let module_ty = module_name.and_then(|module_name| self.module_ty_from_name(module_name));
+        let module_ty = if let Some(module_name) = module_name {
+            if let Some(module_ty) = self.module_ty_from_name(module_name) {
+                module_ty
+            } else {
+                self.unresolved_module_name(import_from, *level as usize, module);
+                Type::Unknown
+            }
+        } else {
+            tracing::debug!("Failed to resolve import due to invalid syntax");
+            Type::Unknown
+        };
 
         let ast::Alias {
             range: _,
             name,
             asname: _,
         } = alias;
 
-        // If a symbol is unbound in the module the symbol was originally defined in,
-        // when we're trying to import the symbol from that module into "our" module,
-        // the runtime error will occur immediately (rather than when the symbol is *used*,
-        // as would be the case for a symbol with type `Unbound`), so it's appropriate to
-        // think of the type of the imported symbol as `Unknown` rather than `Unbound`
-        let member_ty = module_ty
-            .unwrap_or(Type::Unbound)
-            .member(self.db, &Name::new(&name.id))
-            .replace_unbound_with(self.db, Type::Unknown);
+        let member_ty = module_ty.member(self.db, &Name::new(&name.id));
 
-        if matches!(module_ty, Err(ModuleResolutionError::UnresolvedModule)) {
-            self.add_diagnostic(
-                AnyNodeRef::StmtImportFrom(import_from),
-                "unresolved-import",
-                format_args!(
-                    "Import '{}{}' could not be resolved.",
-                    ".".repeat(*level as usize),
-                    module.unwrap_or_default()
-                ),
-            );
-        } else if module_ty.is_ok() && member_ty.is_unknown() {
+        // TODO: What if it's a union where one of the elements is `Unbound`?
+        if member_ty.is_unbound() {
             self.add_diagnostic(
                 AnyNodeRef::Alias(alias),
                 "unresolved-import",
                 format_args!(
-                    "Could not resolve import of '{name}' from '{}{}'",
+                    "Module '{}{}' has no member '{name}'",
```

Comment:
I'm happy with this error message; if you prefer hewing closer to the runtime error and going with `Cannot import '{}' from module '{}'`, I'm fine with that, too.

Reply:
I actually quite like this message. I think it's specifically the term "attribute" that I'd like to avoid here -- although it's accurate, it's also confusing IMO, because people don't _think_ of `from` imports as attribute accesses (even though they are), and because we'll likely have an `attribute-access` or `unknown-attribute` error code that's separate to the `unresolved-import` error code. Referring to a "member" rather than an "attribute" sidesteps this without being less precise (in my opinion), and is also slightly more specific than the runtime error message about _why_ the import failed.